lib/db_connection/connection.ex

```diff
@@ -87,16 +95,28 @@ defmodule DBConnection.Connection do
     else
       {:ok, state} when after_connect != nil ->
         ref = make_ref()
-        Connection.cast(self(), {:after_connect, ref})
-        {:ok, %{s | state: state, client: {ref, :connect}}}
+        :gen_statem.cast(self(), {:after_connect, ref})
+        {:keep_state, %{s | state: state, client: {ref, :connect}}}
 
       {:ok, state} ->
         backoff = backoff && Backoff.reset(backoff)
         ref = make_ref()
-        Connection.cast(self(), {:connected, ref})
-        {:ok, %{s | state: state, client: {ref, :connect}, backoff: backoff}}
+        :gen_statem.cast(self(), {:connected, ref})
+        {:keep_state, %{s | state: state, client: {ref, :connect}, backoff: backoff}}
 
       {:error, err} when is_nil(backoff) ->
+        Logger.error(
+          fn ->
+            [
+              inspect(mod),
+              " (",
+              inspect(self()),
+              ") failed to connect: " | Exception.format_banner(:error, err, [])
+            ]
+          end,
+          crash_reason: {err, []}
+        )
```

Comment:
@josevalim I had to add this here too. From testing with ecto_sql, it turns out that `raise`ing here doesn't log anything for gen_statem, but it did log for Connection. I’m confused, but I also suck at OTP logging, so I’m not surprised 🙃 

If you have better ideas, I’m all ears. 🙃